compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -2277,10 +2277,11 @@ fn confirm_impl_trait_in_trait_candidate<'tcx>(
         obligation.param_env,
         cause.clone(),
         obligation.recursion_depth + 1,
-        tcx.bound_return_position_impl_trait_in_trait_tys(impl_fn_def_id)
-            .map_bound(|tys| {
-                tys.map_or_else(|guar| tcx.ty_error(guar), |tys| tys[&obligation.predicate.def_id])
-            })
+        tcx.collect_return_position_impl_trait_in_trait_tys(impl_fn_def_id)
+            .map_or_else(
+                |guar| ty::EarlyBinder(tcx.ty_error(guar)),
+                |tys| tys.map_bound(|tys| tys[&obligation.predicate.def_id]),
+            )
             .subst(tcx, impl_fn_substs),
```

Comment:
You could probably just push this `subst` call into the "Ok" branch of the `map_or_else`, rather than wrapping the `ty::Error` in a binder.